### PR TITLE
tests: Update DevSim Tests

### PR DESCRIPTION
**Pull request by jeremyk-lunarg**
Updating devsim tests with a new Python test script, new tests for
portability, and new comparison files.

### CODE REVIEW

File: tests/devsim_layer_tests.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+    )
+    
+    args = parser.parse_args()
+    
+    test_dict = {}
+    
+    vulkaninfo_path = args.vulkan_tools_install_dir
+    
+    if sys.platform.startswith("win32"):
+        vulkaninfo_path = os.path.join(vulkaninfo_path, "bin", "vulkaninfo.exe")
```
Comment (lunarpapillo) [inline tests/devsim_layer_tests.py:152]: I'm not a fan of OS-dependent code; but I don't have an easy way to make this sequence OS-independent, so I'm only grumbling about it.

File: tests/devsim_layer_tests.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+    elif sys.platform.startswith("linux"):
+        vulkaninfo_path = os.path.join(vulkaninfo_path, "bin", "vulkaninfo")
+    elif sys.platform.startswith("darwin"):
+        vulkaninfo_path = os.path.join(vulkaninfo_path, "vulkaninfo", "vulkaninfo")
```
Comment (lunarpapillo) [inline tests/devsim_layer_tests.py:156]: Is this correct on Mac?

File: tests/devsim_layer_tests.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    
+    test_dict["BasicTest"] = TestFramework(test_in_files, BasicTest(TestEnvironmentSettings(), "test1_vulkaninfo_out.json", "devsim_test2_gold.json"), vulkaninfo_path)
+    test_dict["PortabilityExtensionPresentEmulationOffTest"] = TestFramework(["devsim_dummy_in.json"], PortabilityExtensionPresentTest(TestEnvironmentSettings(), "port_extent_test_vulkaninfo_out.json", None), vulkaninfo_path)
+    test_dict["PortabilityExtensionPresentEmulationOnTest"] = TestFramework(["devsim_dummy_in.json"], PortabilityExtensionPresentTest(TestEnvironmentSettings(portability=1), "port_extent_test_vulkaninfo_out.json", None), vulkaninfo_path)
+    test_dict["PortabilityNonEmulatedTest"] = TestFramework(["portability_test.json"], PortabilityTest(TestEnvironmentSettings(), "port_test_vulkaninfo_out.json", "portability_test_gold.json"), vulkaninfo_path)
+    test_dict["PortabilityEmulatedTest"] = TestFramework(["portability_test.json"], PortabilityTest(TestEnvironmentSettings(portability=1), "port_test_vulkaninfo_out.json", "portability_test_gold.json"), vulkaninfo_path)
```
Comment (lunarpapillo) [inline tests/devsim_layer_tests.py:164]: As above, it's easier and clearer to:
```
test_dict = {
    "BasicTest": TestFramework(...),
    "PortabilityExtensionPresentEmulationOffTest": TestFramework(...),
    ...
}
```

File: tests/devsim_layer_tests.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+    
+    if not test_dict["PortabilityExtensionPresentEmulationOffTest"].runAndCheck():
+        test_blacklist.append("PortabilityExtensionPresentEmulationOffTest")
+        test_blacklist.append("PortabilityNonEmulatedTest")
+    
+    for k in test_dict:
+        if k in test_blacklist:
+            continue
+        print("RUNNING TEST")
+        print(k)
```
Comment (lunarpapillo) [inline tests/devsim_layer_tests.py:176]: Suggest: `print("RUNNING TEST", k)`

File: tests/devsim_layer_tests.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+        print("{}".format("PASS" if test_dict[k].runAndCheck() else "FAIL"))
```
Comment (lunarpapillo) [inline tests/devsim_layer_tests.py:177]: Don't need the `"{}".format()` construction.  Suggest: 
```
print("PASS" if text_dict[k].runAndCheck() else "FAIL")
```

File: tests/devsim_test2_gold.json [hunk 6]
```diff
@@ -58,61 +32,61 @@
...
-    "sparseResidencyImage2D": 147,
-    "sparseResidencyImage3D": 148,
-    "tessellationShader": 106,
-    "textureCompressionASTC_LDR": 122,
-    "textureCompressionBC": 123,
-    "textureCompressionETC2": 121,
-    "variableMultisampleRate": 154,
-    "vertexPipelineStoresAndAtomics": 126,
-    "wideLines": 116
+    "alphaToOne": 1,
```
Comment (lunarpapillo) [inline tests/devsim_test2_gold.json:35]: Were the old values IDs of some kind?  I don't understand why these were changed.

File: tests/portability_test.json [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+    },
+    "VkPhysicalDevicePortabilitySubsetFeaturesKHR": {
+        "imageViewFormatReinterpretation": 0,
+        "pointPolygons": 0,
+        "samplerMipLodBias": 0
+    },
+    "VkPhysicalDevicePortabilitySubsetPropertiesKHR": {
+        "minVertexInputBindingStrideAlignment": 8
+    }
+}
```
Comment (lunarpapillo) [inline tests/portability_test.json:17]: Seems to be missing a terminal newline, where almost all the other files have one.

File: tests/portability_test_gold.json [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+		"shaderSampleRateInterpolationFunctions": 1,
+		"tessellationIsolines": 1,
+		"tessellationPointMode": 1,
+		"triangleFans": 1,
+		"vertexAttributeAccessBeyondStride": 1
+    },
+    "VkPhysicalDevicePortabilitySubsetPropertiesKHR": {
+        "minVertexInputBindingStrideAlignment": 8
+    }
+}
```
Comment (lunarpapillo) [inline tests/portability_test_gold.json:22]: Seems to be missing a termianl newline, where almost all of the other files have one.

**Review by lunarpapillo — CHANGES_REQUESTED** [review]
Lots of good work here!  So I have lots of suggestions. :-)

It's not clear what version of Python is being targeted here.  Whatever version is being used, the corresponding version of `pylint` should also be used to validate the Python.  (`pylint` will catch many of the things I caught, and more.)

It's not clear to me where these tests are going to be run.  Because they are in the **VulkanTools** repository, I presume these are developer tests and/or Travis/AppVeyor tests?  If they're intended to run in internal LunarG CI, there's a lot more that needs to be done, to match the internal test harness.

The commit message is insufficient to justify all the changes here (in particular, the changes to values in the `.json` files) - some description of why that was done should appear.

**Comment by jeremyk-lunarg** [comment]
@lunarpapillo, I am intending for this to be used by internal CI.

What additional changes do I need to make for it to be compatible with the internal test harness?

**Comment by lunarpapillo** [comment]
> @lunarpapillo, I am intending for this to be used by internal CI.
> 
> What additional changes do I need to make for it to be compatible with the internal test harness?

Internal tests all live in **LunarG/VulkanTests**, and use the utilities and test classes there, which provide a lot of features, including:
-  full logging, to help determine what happened when a test failed
- capture of all stdout and stderr for all external operations
- saved test artifacts/results for later review
- OS-independent operation, requiring all OS-dependent operations to be abstracted out and placed in OS-dependent classes, to allow a test to run on all supported OSes without having to change if a new OS is added
- etc...

The options here are:
- move test scripts entirely into **VulkanTests**
- create a new **VulkanTests** script that invokes the project scripts as a module
- create a new **VulkanTests** script that invokes the project scripts as an external executable

The choice depends on whether the test scripts are intended to be available to developers, Travis, or AppVeyor; or whether they're just intended for LunarG internal CI.

For more on the test harness: https://github.com/LunarG/VulkanTests/wiki/Test-Suite-Architecture

File: tests/devsim_test2_gold.json (hunk 6 above)
Comment (jeremyk-lunarg) [reply tests/devsim_test2_gold.json:35]: These values are `VkBool32` which can technically be any 32-bit integer.

The original test made these values weird, ridiculous numbers to show that devsim could simulate a driver with values that didn't make any sense.

Due to updates in `vulkaninfo`, these values are forced to `0` or `1` in output, so in order for the tests to pass now, the gold files had to be updated to the new `vulkaninfo` output.

**Comment by lunarpapillo** [comment]
Woot!  Looks good, but I didn't review the changes in `devsim_layer_tests.py` closely, as they look similar/identical to what they were during our live review.

Did you have to change this file much?  If you could direct me to the locations that changed, I'll review those more thoroughly.

Also, it looks like your Ubuntu builds are failing, likely because a file is missing from this commit...

**Comment by jeremyk-lunarg** [comment]
@lunarpapillo It's passing now!

The changes I made to `devsim_layer_tests.py` are in commit 18bd6e34b15347ed27342bd308e2b1284910c145.

In summary:
- I had to add another flag that was added to Devsim recently.
- I fixed a test that had improper settings.
- I removed memory flag modification from one of the tests.  This will be tested in a future test.

I also made some changes in commit ab3abca5191b9d6583615afd3f699427b602e11d.

In summary:
- The tests need to be run in the same directory as the test files.  I added an argument to allow a user to set the run directory from the terminal.

It was a quick change to make the GitHub Actions work and could be made cleaner if this is going to be used as a module, but I'll let you be the judge of that.

File: tests/devsim_layer_tests.py [hunk 9]
```diff
@@ -363,10 +363,22 @@ def main():
...
 
     args = parser.parse_args()
 
     vulkaninfo_path = args.vulkan_tools_install_dir
+    working_directory = args.directory
+
+    starting_working_directory = os.getcwd()
+
+    if working_directory is not None:
+        working_directory = os.path.abspath(working_directory)
```
Comment (lunarpapillo) [inline tests/devsim_layer_tests.py:380]: It's easier to provide the value as a default argument:
```
parser.add_argument(
    "--dir",
    help="Specify the directory to run the tests in.  Defaults to current working directory.",
    dest="directory",
    default=os.getcwd()
)
```
Then after `parse_args()`:
```
working_directory = os.path.abspath(args.directory)
```

**Review by lunarpapillo — APPROVED** [review]
Looks great!  I suggest one change, but the code is clear and understandable as is, so if you don't want to make the suggested change I still approve. :+1:

File: tests/devsim_layer_tests.py (hunk 9 above)
Comment (jeremyk-lunarg) [reply tests/devsim_layer_tests.py]: Good suggestion!
Made the change and pushed it up.  Will merge after cloud CI passes.
